### PR TITLE
Added how-to-implement-structured-logging-in-adonisjs

### LINKED ISSUE #2442: How to implement Structured Logging in Adonisjs
### How-To Guides Overview 
I will be showing you the best practice in implementing Structured in Adonisjs, discussing in detail the best ways involved in implementing a scalable Logging System in your RESTFUL API. We will build a real-world app to demonstrate how to debug easily with a good Structured Log in AdonisJS.

### What is the Scope and Purpose of the Guide
Students will learn how to build a high-maintainable app with Adonisjs
Students will learn how to implement a Structured Logging in AdonisJS Restful API
Students will learn how to choose what to Log and how to Log it.
Students will learn about the concept of Structured Logging and how to maintain a large App with it.
Students will learn best practices in building large-scale Web apps with AdonisJS.

References:
https://www.scalyr.com/blog/the-10-commandments-of-logging/
https://stackify.com/what-is-structured-logging-and-why-developers-need-it/
https://www.loggly.com/blog/30-best-practices-logging-scale/

### PR DESCRIPTION
**Pull request by Kaperskyguru**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This fixes #2442

### CODE REVIEW

**Comment by Kaperskyguru** [comment]
Hello @geoffrey45,

Been a while. I will really love it if you could take a look at this article.

Thanks.

**Comment by geoffrey45** [comment]
@Kaperskyguru I will start the review once I'm done with the one before you. 🚀🚀

File: content/articles/how-to-implement-structured-logging-in-adonisjs/index.md [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+![Log Errors](/engineering-education/how-to-implement-structured-logging-in-adonisjs/log_errors.png)
+
+Next, we will update our Forum API to log out specific and vital information of our application as the user interacts with it in production.
+
+First, we will update the `ForumsController.ts` in `app/Controllers/Http`:
+
+```js
+import { HttpContextContract } from "@ioc:Adonis/Core/HttpContext";
+import Forum from "App/Models/Forum";
+import Logger from "@ioc:Adonis/Core/Logger";
```
Comment (geoffrey45) [inline content/articles/how-to-implement-structured-logging-in-adonisjs/index.md:98]: For all the code blocks:
- Try keeping them as short as possible. 
- Don't add unused code. No repetition.
- Break it up into chunks for the reader to follow along. 
- Explain what this file does, why we are changing it, and what to expect after applying the changes.
- Don't put all the changes for a single file in a single block. Changes in different locations in the file should be discussed individually, with clear explanation and instructions.
- All of the above should be in a paragraph just above the code block.

Please apply the above to all your code blocks.

File: content/articles/how-to-implement-structured-logging-in-adonisjs/index.md [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        .delete();
+    Logger.info({ UserID: user.id }, `Forum deleted: ${forum}`);
+    await Cache.delete("forum_id_" + params.id);
+    return 404;
+    }
+}
+```
+
+In the code above, we have added structured Logging to the controller. The controller logs general information, but specific message and the IDs of when it happens.
+
```
Comment (geoffrey45) [inline content/articles/how-to-implement-structured-logging-in-adonisjs/index.md:196]: Provide granular explanation. State what, where and why exactly you added something. For example, using a generic example:

- "We have created a new array `oddNumber` to hold all the odd numbers. This array will be updated after every iteration".

**Review by geoffrey45 — CHANGES_REQUESTED** [review]
@Kaperskyguru Thank you for submitting your article for review. Great article. 

Please check the attached comments. Leave a mention in a comment when you are done with the requested changes.

**Comment by hectorkambow** [comment]
plagiarism check done


Please see report - a couple minor instances that may require attention. 

[how-to-implement-structured-logging-in-adonisjs #2973.pdf](https://github.com/section-engineering-education/engineering-education/files/6895753/how-to-implement-structured-logging-in-adonisjs.2973.pdf)

**Comment by Kaperskyguru** [comment]
Thanks. I will rework this including the pointers @geoffrey45 listed out.

**Review by Kaperskyguru — COMMENTED** [review]
Awesome. Thanks

File: content/articles/how-to-implement-structured-logging-in-adonisjs/index.md [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+Getting started with Adonis.js is easy because of its good documentation. To learn more about the latest Adonis.js 5, reading through [Adonis.js 5 ultimate guide](https://masteringbackend.com/posts/adonisjs-tutorial-the-ultimate-guide) and [building a RESTFUL API with Adonis 5](/engineering-education/build-a-restful-api-with-adonisjs/) to get up to speed.
+
+To install Adonis.js 5, run the command below:
+
+```bash
+npm init adonis-ts-app@latest forum-api
+```
+
+The command creates a new Adonis.js 5 project, properly configured and ready to be used. We will continue with the Forum API developed with the article on: [How to Implement Caching using Adonis.js 5 tutorial](/engineering-education/how-to-implement-caching-in-adonisjs-5/). You can [clone the repository](https://github.com/Kaperskyguru/adonisjs-forum-api) for an easy setup.
+
```
Comment (geoffrey45) [inline content/articles/how-to-implement-structured-logging-in-adonisjs/index.md:41]: Does the reader have to read the "How to Implement Caching using Adonis.js 5" to be able to follow along with this tutorial? 

It is very hard to follow along with this tutorial, because:
1. The Forum app is an already built app. And:
2. You have not provided information on how to set it up. 
3. You have not explained what this forum app does.
4. There are several libraries which you are using already. ie. Adonis.js, Pino.js, which make it even more confusing.
5. The amount of code in this tutorial is very huge and not well documented and explained.

So, the readers are editing files that contain existing code. Which they don't know what the code does.

Please consider setting up another simple project from scratch to demonstrate the concept you are demonstrating in this tutorial.

File: content/articles/how-to-implement-structured-logging-in-adonisjs/index.md [hunk 4]
```diff
@@ -0,0 +1,261 @@
+## How to implement Structured Logging in Adonis.js
+Logs are among the essential tools for application developers during the debugging process and for analysts monitoring applications in the production environment.
+
+Typical logs are textual formats that are easily readable by humans but very difficult for machines to read and process.
+
+Assuming you want to use automated processing to investigate log files, or you want to use algorithms to categorize, index, and search through log files based on specific parameters, typical logs will not work as expected.
+
+That is where structured logging becomes useful. Structured logging is a way to record the information in a way that is easy for machines to process.
+
+This article will show the best practice in implementing structured logging in Adonis.js 5, discussing in detail, the best ways involved in implementing a scalable logging system in your RESTFUL API.
```
Comment (geoffrey45) [inline content/articles/how-to-implement-structured-logging-in-adonisjs/index.md:10]: Here are some of the major issues I encountered while going through a prelim review on your article.

1. Take note of your spelling. Be consistent. If you choose to use for example the word "Adonis.js", you should use it consistently throughout your tutorial. Don't interchange it with "Adonis" or any other spelling of the word. eg. "AdonisJS".
2. Inconsistent capitalization. There were many case of capitalized word in a sentence. For example, the words "structured logging" were most of the time capitalized in sentense. ie. "Structured Logging". Please see [this thesaurus page](https://www.thesaurus.com/e/grammar/when-to-capitalize-words/) on capitalization rules.


> 🔷 **Please take note of the above in future tutorials.**

**Review by geoffrey45 — CHANGES_REQUESTED** [review]
@Kaperskyguru Please see the attached comments.

File: content/articles/how-to-implement-structured-logging-in-adonisjs/index.md [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+Logs are among essential tools for application developers during the debugging process and for analysts monitoring applications in the production environment.
+
+Typical logs are textual formats that are easily readable by humans but very difficult for machines to read and process.
+
+Assuming you want to use automated processing to investigate log files, or you want to use algorithms to categorize, index, and search through log files based on specific parameters, typical logs will not work as expected.
+
+That is where Structured Logging becomes useful:
+
+This article will show the best practice in implementing Structured Logging in Adonisjs, discussing in detail the best ways involved in implementing a scalable Logging System in your RESTFUL API. We will build a real-world app to demonstrate how to debug easily with an excellent Structured Log in AdonisJS.
+
```
[the remark this answers is not in the document]
Comment (Kaperskyguru) [reply content/articles/how-to-implement-structured-logging-in-adonisjs/index.md]: Done

File: content/articles/how-to-implement-structured-logging-in-adonisjs/index.md [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+I will be showing you the best practice in handling Structured Logging in Adonisjs, and we will build a real-world Forum app to demonstrate how to boost maintainability in AdonisJS.
+
+In this tutorial, you will learn:
+
+- How to build a high-maintainable app with Adonisjs.
+- Implementing a Structured Logging in AdonisJS Restful API.
+- How to choose what to Log and how to Log it.
+- Learn about the concept of Structured Logging and how to maintain a large App with it.
+- Learn best practices in building large-scale Web apps with AdonisJS.
```
[the remark this answers is not in the document]
Comment (Kaperskyguru) [reply content/articles/how-to-implement-structured-logging-in-adonisjs/index.md]: Done

File: content/articles/how-to-implement-structured-logging-in-adonisjs/index.md [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+Getting started with Adonis 5 is easy because of its well-documented documentation. To learn more about the latest AdonisJS 5, reading through [Adonis.js 5 ultimate guide](https://masteringbackend.com/posts/adonisjs-tutorial-the-ultimate-guide) and [building a RESTFUL API with Adonis 5](/engineering-education/build-a-restful-api-with-adonisjs/) to get up to speed.
+
+To install AdonisJS 5, run the command below:
+
+```bash
+npm init adonis-ts-app@latest forum-api
+```
+
+The command creates a new Adonis 5 project, properly configured and ready to be used. We will continue with the Forum API developed with the article on [How to Implement Caching using Adonis.js 5](/engineering-education/how-to-implement-caching-in-adonisjs-5/). You can [clone the repository](https://github.com/Kaperskyguru/adonisjs-forum-api) for an easy setup.
+
```
[the remark this answers is not in the document]
Comment (Kaperskyguru) [reply content/articles/how-to-implement-structured-logging-in-adonisjs/index.md]: Done

File: content/articles/how-to-implement-structured-logging-in-adonisjs/index.md (hunk 3 above)
Comment (Kaperskyguru) [reply content/articles/how-to-implement-structured-logging-in-adonisjs/index.md:41]: The article is more of a series. It follows along with 2 articles I have already written

**Comment by Kaperskyguru** [comment]
Hello @geoffrey45,

So sorry for delays. I have been on vacation for a while.

I have fixed all the issues raised. Please go through again.

Thanks.

**Comment by ahmadmardeni1** [comment]
Hey, @Kaperskyguru👋 Great tutorial!
However, as @geoffrey45 said, these two concepts ([cashing](https://www.section.io/engineering-education/how-to-implement-caching-in-adonisjs-5/) and logging) can exist independently of each other. Thus, you don't have to do caching to do structured logging. But you will be forced in this article since the code is from the caching article.

I will close this PR simply to clear the queue. Please set up a new project from scratch for this tutorial, @ me, and I will reopen the PR.

If you need help or you have any questions, please reach me or your peer reviewer via Slack.
